### PR TITLE
Override evaluation

### PR DESCRIPTION
**Pull request by mehalshah**
After taking the evaluation, users now get directed to a multiselect box that lets them pick other modules if they don't like what they are assigned.

### CODE REVIEW

**Comment by mehalshah** [comment]
I made an interstitial "Preview your learning modules page" that doesn't actually enroll you in anything until you hit the confirm button - that's what actually creates your enrollments.

UI is not final, there's a separate task this sprint for that work

**Comment by joshlory** [comment]
Missing change in `enrollment_unit_assignment.rb`?

```
NameError at /plc/user_course_enrollments
=========================================

> uninitialized constant Plc::EnrollmentUnitAssignment::REVIEWING_ASSIGNMENTS
```

File: code-studio/src/js/plc/override_evaluation.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+    $('#learning_module_ids option:selected').each( function () {
+      if (this.value) {
+        moduleAssignments.push(this.value);
+      }
+    });
+
+    $('#learning_module_list').val(moduleAssignments);
+  });
+});
```
Comment (joshlory) [inline code-studio/src/js/plc/override_evaluation.js]: Not used?

File: code-studio/scripts/build-js.js [hunk 2]
```diff
@@ -70,8 +70,9 @@ Promise.all([
   // Have a bundle for plc stuff - no sense in expanding this to everything yet
   build_commands.bundle(_.extend({}, defaultOptions, {
     filenames: [
-      'plc/perform_evaluation.js',
       'plc/evaluation_creation.js',
+      'plc/override_evaluation.js',
```
Comment (joshlory) [inline code-studio/scripts/build-js.js]: Not used?

**Comment by joshlory** [comment]
We should add a test case to cover the `REVIEWING_ASSIGNMENTS` break above.
- Added

File: dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+You will be assigned these modules
+- @enrolled_modules.each do |enrolled_module|
+  %li
+    = enrolled_module.name
+
+%br
+You can change your course enrollment here
+= form_tag(action: 'confirm_assignments') do
+  - options = course_unit.plc_learning_modules.pluck(:name, :required, :id).map {|name, required, id| ["#{name}#{' - Required' if required}", id]}
```
Comment (joshlory) [inline dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml]: Let's move this logic out of the view, and combine with the other `' - Required'` logic if possible.

I'm still a fan of providing a method on the `LearningModule` model that gets the name with `' - Required'` if the module is required.

File: dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  = select_tag 'learning_module_ids', options_for_select(options, selected: @enrolled_modules.pluck(:id), disabled: course_unit.plc_learning_modules.where(required: true).pluck(:id)), {multiple: true, size: [options.size, 25].min, style: 'width: 100%'}
```
Comment (joshlory) [inline dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml]: Anytime you're writing Active Record queries in a view there's probably a better way to do it 😄 .  Let's move this to the model (that uses a scope of required learning modules).

Comment (joshlory) [inline dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml]: This is the second usage of the `size: [options.size, 25].min` logic which makes this a good candidate for extracting a `max_options_select_tag` helper.  See [`boolean_check_box`](https://github.com/code-dot-org/code-dot-org/blob/d29dcf90cf7654371dcede0f39c1d622d9f5bfa6/dashboard/app/helpers/levels_helper.rb#L491-L493) for a good example.

File: dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+  = hidden_field_tag('learning_module_list')
```
Comment (joshlory) [inline dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml]: Not used?

File: dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml (hunk 3 above)
Comment (mehalshah) [inline dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml]: So I have a change in flight to move required from a boolean column to a column called module_type (required, content, practice). I'll implement that method when I do that.

File: dashboard/app/models/plc/enrollment_unit_assignment.rb [hunk 6]
```diff
@@ -40,6 +40,8 @@ def enroll_user_in_unit_with_learning_modules(learning_modules)
       learning_modules |=  self.plc_course_unit.plc_learning_modules.where(required: true)
 
       learning_modules.each do |learning_module|
+        next unless learning_module.plc_course_unit == self.plc_course_unit
```
Comment (joshlory) [inline dashboard/app/models/plc/enrollment_unit_assignment.rb:44]: Why is this needed?  (Maybe add a comment explaining...)

File: dashboard/app/controllers/plc/enrollment_evaluations_controller.rb [hunk 7]
```diff
@@ -1,25 +1,44 @@
...
+    # Eventually, the logic for more complex module assignments will go here. For now since we are naively assigning modules, just go to preview
+    redirect_to controller: :enrollment_evaluations, action: :preview_assignments, unit_assignment_id: params[:unit_assignment_id], enrolled_modules: params[:answer_module_list]
+  end
+
+  def preview_assignments
+    @enrollment_unit_assignment = Plc::EnrollmentUnitAssignment.find(params[:unit_assignment_id])
+    @enrolled_modules = Plc::LearningModule.where(id: params[:enrolled_modules].split(','))
+
+    if @enrolled_modules.where.not(plc_course_unit: @enrollment_unit_assignment.plc_course_unit).any?
+      redirect_to controller: :enrollment_evaluations, action: :perform_evaluation, unit_assignment_id: @enrollment_unit_assignment.id
```
Comment (joshlory) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb]: Why is this needed?  (Maybe add a comment explaining...)

File: dashboard/app/controllers/plc/enrollment_evaluations_controller.rb [hunk 8]
```diff
@@ -1,25 +1,44 @@
...
+
+  private
+  def authorize_and_check_unit_state
+    plc_unit_assignment = Plc::EnrollmentUnitAssignment.find(params[:unit_assignment_id])
+    authorize! :read, plc_unit_assignment.plc_user_course_enrollment
 
-    enrollment_unit_assignment.enroll_user_in_unit_with_learning_modules(modules_to_enroll_in)
-    enrollment_unit_assignment.update(status: Plc::EnrollmentUnitAssignment::IN_PROGRESS)
-    redirect_to controller: :enrollment_unit_assignments, action: :show, id: enrollment_unit_assignment.id
+    raise 'Cannot do evaluation actions when unit is in progress' unless Plc::EnrollmentUnitAssignment::PENDING_EVALUATION == plc_unit_assignment.status
```
Comment (joshlory) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb]: Clicking the back button after confirming module selections throws this error.

File: dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml (hunk 3 above)
Comment (joshlory) [inline dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml]: Can we discuss this a little more?  A module could be of type "content" and also be "required".

Comment (joshlory) [inline dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml]: Followed up offline — I'm wrong, these are all distinct things 😄 .

I think we can still combine the logic now?  Then when you change the column you'll only need to update one place instead of two.

**Comment by joshlory** [comment]
Done w/ my first pass.

File: code-studio/src/js/plc/override_evaluation.js (hunk 1 above)
Comment (mehalshah) [inline code-studio/src/js/plc/override_evaluation.js]: Removed.

File: dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml (hunk 5 above)
Comment (mehalshah) [inline dashboard/app/views/plc/enrollment_evaluations/preview_assignments.html.haml]: Removed

File: dashboard/app/models/plc/enrollment_unit_assignment.rb (hunk 6 above)
Comment (mehalshah) [inline dashboard/app/models/plc/enrollment_unit_assignment.rb:44]: I don't want someone to post to the controller with an arbitrary list of learning module ids that might be from various sources. This is to ensure that all the learning modules we are enrolling you in are in this course unit.

I'll add a comment

File: dashboard/app/controllers/plc/enrollment_evaluations_controller.rb (hunk 7 above)
Comment (mehalshah) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb]: Yeah - its not necessary right now. I'll take it out

File: dashboard/app/controllers/plc/enrollment_evaluations_controller.rb (hunk 8 above)
Comment (mehalshah) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb]: When you first confirmed your module selections, then the unit was marked as in progress. Going back doesn't undo that, so yeah you should get this error.

Right now, we don't want users to be able to retake the evaluation or change their module selection once they've picked their units. Future iterations on this may allow you to change your enrollment if you haven't started task completion work

Comment (mehalshah) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb]: But if its a pain to test / debug, I don't mind taking this out

Comment (joshlory) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb]: That's fair, but raising a 500 error (which looks like the image below) probably isn't the most user-friendly way to convey this info.  Maybe show an [error message flash](https://github.com/code-dot-org/code-dot-org/blob/ef809a22b21a0aa9da8b9b43967be89178f5bc9b/dashboard/app/controllers/sections_controller.rb#L31)?

Comment (mehalshah) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb]: I'm just going to take this out - it makes testing a little annoying

File: dashboard/app/models/plc/course_unit.rb [hunk 9]
```diff
@@ -26,4 +26,8 @@ class Plc::CourseUnit < ActiveRecord::Base
   def get_all_possible_learning_resources
     plc_learning_modules.map(&:plc_tasks).flatten.select{ |task| task.class == Plc::LearningResourceTask }
   end
+
+  def get_required_learning_module_ids
+    plc_learning_modules.where(required: true).pluck(:id)
```
Comment (joshlory) [inline dashboard/app/models/plc/course_unit.rb]: From my previous feedback, required learning modules probably makes sense as a scope.

File: dashboard/app/helpers/plc/learning_modules_helper.rb [hunk 10]
```diff
@@ -3,10 +3,13 @@ def options_for_plc_task_learning_modules
...
-      ["#{name}#{' - Required' if required}", id]
+  def options_with_learning_module_names_and_required_tag(course_unit)
+    course_unit.plc_learning_modules.order(:required, :name).map do |learning_module|
+      [learning_module.get_name_with_required_tag, learning_module.id]
     end
   end
+
+  def learning_modules_select_size options
+    [options.size, 25].min
+  end
```
Comment (joshlory) [inline dashboard/app/helpers/plc/learning_modules_helper.rb:14]: Nit: I'm not sure how much of a win it is to extract `[options.size, 25].min` to a helper (or why passing `options` is preferable to passing `size`).

From my earlier feedback, prefer a `max_options_select_tag` helper like [`boolean_check_box`](https://github.com/code-dot-org/code-dot-org/blob/d29dcf90cf7654371dcede0f39c1d622d9f5bfa6/dashboard/app/helpers/levels_helper.rb#L491-L493).  Or better yet, maybe a LearningModule-specific select tag helper.

I think we can further de-dupe these similar tags:

``` ruby
= select_tag 'learning_module_ids', options_for_select(options, selected: @enrolled_modules.pluck(:id), disabled: course_unit.get_required_learning_module_ids), {multiple: true, size: learning_modules_select_size(options), style: 'width: 100%'}
```

``` ruby
= f.select :plc_learning_module_ids, options_for_select(options, @task.plc_learning_modules.pluck(:id)), {}, {multiple: true, size: learning_modules_select_size(options), style: 'width: 100%'}
```

``` ruby
= f.select :plc_learning_module_ids, options_for_select(options, @task.plc_learning_modules.map(&:id)), {}, {multiple: true, size: [options.size, 25].min, style: 'width: 100%'}
```

File: dashboard/app/controllers/plc/enrollment_evaluations_controller.rb (hunk 7 above)
Comment (joshlory) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb]: Still to be removed?

File: dashboard/app/controllers/plc/enrollment_evaluations_controller.rb [hunk 11]
```diff
@@ -1,25 +1,40 @@
...
+
+    if @enrolled_modules.where.not(plc_course_unit: @enrollment_unit_assignment.plc_course_unit).any?
+      redirect_to controller: :enrollment_evaluations, action: :perform_evaluation, unit_assignment_id: @enrollment_unit_assignment.id
+    end
+  end
 
-    modules_to_enroll_in = Plc::LearningModule.where(id: question_responses)
+  def confirm_assignments
+    modules_to_enroll_in = Plc::LearningModule.where(id: params[:learning_module_ids])
+    enrollment_unit_assignment = Plc::EnrollmentUnitAssignment.find(params[:unit_assignment_id])
```
Comment (joshlory) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb:23]: Can any of this be pushed to the model?  We're doing a lot of ActiveRecord work in the controller action.

**Comment by joshlory** [comment]
Added a few more minor feedback items.  The TravisCI failure looks legit.

File: dashboard/app/controllers/plc/enrollment_evaluations_controller.rb (hunk 11 above)
Comment (mehalshah) [inline dashboard/app/controllers/plc/enrollment_evaluations_controller.rb:23]: I have to get the enrollment unit assignment here - all I have is the enrollment unit assignment ID at this stage, not the object.

The modules to enroll in can get moved to model if I make the call take a list of module ids instead of modules. How large is the benefit?

File: dashboard/app/helpers/plc/learning_modules_helper.rb (hunk 10 above)
Comment (mehalshah) [inline dashboard/app/helpers/plc/learning_modules_helper.rb:14]: Unifying the last two is doable. Unifying the first one with them is problematic, since it's not part of a form for an object

Comment (mehalshah) [inline dashboard/app/helpers/plc/learning_modules_helper.rb:14]: Think I can just do this with form_tag_helper

Comment (mehalshah) [inline dashboard/app/helpers/plc/learning_modules_helper.rb:14]: This refactor got way too ugly.

**Comment by mehalshah** [comment]
Did some of the PR feedback, but I don't understand the justification for the last piece

File: dashboard/app/views/plc/tasks/_form.html.haml [hunk 12]
```diff
@@ -11,7 +11,7 @@
     = f.text_field :name
     = f.label :learning_module_id
     - options = options_for_plc_task_learning_modules
-    = f.select :plc_learning_module_ids, options_for_select(options, @task.plc_learning_modules.map(&:id)), {}, {multiple: true, size: [options.size, 25].min, style: 'width: 100%'}
+    = learning_modules_for_task_multi_select(f, options, @task.plc_learning_modules)
```
Comment (joshlory) [inline dashboard/app/views/plc/tasks/_form.html.haml:14]: The `.try(:pluck, :id)` won't get the value here since it hasn't been saved to the DB yet.

File: dashboard/app/views/plc/learning_modules/show.html.haml [hunk 13]
```diff
@@ -3,7 +3,7 @@
 
 %p
   %strong Name:
-  = @learning_module.get_name_with_required_tag
+  = @learning_module.name_with_required_tag
```
Comment (joshlory) [inline dashboard/app/views/plc/learning_modules/show.html.haml:6]: Is there a test that should've caught this?

File: dashboard/app/views/plc/tasks/_form.html.haml (hunk 12 above)
Comment (mehalshah) [inline dashboard/app/views/plc/tasks/_form.html.haml:14]: hmm - since this is the form for new task generation (it's not used for editing) I really shouldn't be passing anything here. Nothing in the box should be selected.

**Comment by joshlory** [comment]
LGTM

File: dashboard/app/views/plc/tasks/_form.html.haml (hunk 12 above)
Comment (joshlory) [inline dashboard/app/views/plc/tasks/_form.html.haml:14]: Oh I thought it could receive an ID from query params?

(On links coming for a specific module, that module used to be selected from the list.)

File: dashboard/app/views/plc/learning_modules/show.html.haml (hunk 13 above)
Comment (mehalshah) [inline dashboard/app/views/plc/learning_modules/show.html.haml:6]: I didn't build UTs for the learning modules controller at first because there wasn't any real logic there beyond basic CRUD. But as its increased in complexity, UTs seem justified at this point. Something that hits the basic functionality will catch this.
